src/Compilers/Server/VBCSCompiler/ServerDispatcher.cs

```diff
@@ -102,20 +102,29 @@ public void ListenAndDispatchConnections(TimeSpan? keepAlive, CancellationToken
                     _clientConnectionHost.EndListening();
                 }
 
-                // This type is responsible for cleaning up resources associated with _listenTask. Once EndListening
-                // is complete this task is guaranteed to be completed. If it ran to completion we need to 
-                // dispose of the value.
-                Console.WriteLine(_listenTask?.Status);
-                Debug.Assert(_listenTask is null || _listenTask.IsCompleted);
-                if (_listenTask?.Status == TaskStatus.RanToCompletion)
+                if (_listenTask is not null)
                 {
-                    try
+                    // This type is responsible for cleaning up resources associated with _listenTask. Once EndListening
+                    // is complete this task is guaranteed to be either completed or have a task scheduled to complete
+                    // it. If it ran to completion we need to dispose of the value.
+                    Console.WriteLine(_listenTask.Status);
+                    if (!_listenTask.IsCompleted)
                     {
-                        _listenTask.Result.Dispose();
+                        // Wait for the task to complete
+                        _listenTask.ContinueWith(_ => { }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default)
+                            .Wait(CancellationToken.None);
```

Comment:
This is adding a sync wait to a place wasn't designed for it. Rather than doing this I think it would help to have a way to check when an `AsyncQueue<T>` has completed all of it's tasks. Just as we could before this change.

Reply:
> This is adding a sync wait to a place wasn't designed for it.

The call to `EndListening` previously inlined part of these operations, but it was the same work getting done at the same point in time.

> ... check when an `AsyncQueue<T>` has completed all of it's tasks ... Just as we could before this change.

This was often true, but there were several conditions under which this was not previously true. Implementing it as described (without race conditions) would be borderline impossible, so I'm instead looking at alternatives:

1. Make `EndListening` and this method asynchronous, so it can `await` instead of block.
2. Keep the code in the current form.
3. Instead of synchronously disposing of objects before `ListenAndDispatchConnections` returns, we just make sure a continuation is registered which would dispose of the object at the right time. If the ordering of dispose operations matters, we can use `ReferenceCountedDisposable<T>` and make sure that each listen task has a lease on the underlying shared disposable object so it can be disposed automatically when all the dependents complete.

Reply:
For now I kept the current approach since it most closely matches the behavior prior to my change.

Reply:
My instinct is to make this method `async`. Unfortunately it's decently important, on Linux at least, that we fully dispose of our named pipe objects before leaving this method (as best as possible). That is likely to be a bit of a ball of yarn to pull on though. 

Maybe leave with current approach for now, file a bug to have me think about this .